addons/stock_account/models/account_move.py

```diff
@@ -1,12 +1,227 @@
+        if self._context.get('move_reverse_cancel'):
+            return super(AccountMove, self).post()
+
+        # Create additional COGS lines for customer invoices.
+        self.env['account.move.line'].create(self._stock_account_prepare_anglo_saxon_out_lines_vals())
+
+        # Post entries.
+        res = super(AccountMove, self).post()
+
+        # Reconcile COGS lines in case of anglo-saxon accounting with perpetual valuation.
+        self._stock_account_anglo_saxon_reconcile_valuation()
+        return res
+
+    @api.multi
+    def button_cancel(self):
+        # OVERRIDE
+        res = super(AccountMove, self).button_cancel()
+
+        # Unlink the COGS lines generated during the 'post' method.
+        self.mapped('line_ids').filtered(lambda line: line.is_anglo_saxon_line).unlink()
+        return res
+
+    # -------------------------------------------------------------------------
+    # COGS METHODS
+    # -------------------------------------------------------------------------
+
+    @api.multi
+    def _stock_account_prepare_anglo_saxon_out_lines_vals(self):
+        ''' Prepare values used to create the journal items (account.move.line) corresponding to the Cost of Good Sold
+        lines (COGS) for customer invoices.
+
+        Example:
+
+        Buy a product having a cost of 9 being a storable product and having a perpetual valuation in FIFO.
+        Sell this product at a price of 10. The customer invoice's journal entries looks like:
+
+        Account                                     | Debit | Credit
+        ---------------------------------------------------------------
+        200000 Product Sales                        |       | 10.0
+        ---------------------------------------------------------------
+        101200 Account Receivable                   | 10.0  |
+        ---------------------------------------------------------------
+
+        This method computes values used to make two additional journal items:
+
+        ---------------------------------------------------------------
+        220000 Expenses                             | 9.0   |
+        ---------------------------------------------------------------
+        101130 Stock Interim Account (Delivered)    |       | 9.0
+        ---------------------------------------------------------------
+
+        Note: COGS are only generated for customer invoices except refund made to cancel an invoice.
+
+        :return: A list of Python dictionary to be passed to env['account.move.line'].create.
+        '''
+        lines_vals_list = []
+        for move in self:
+            if move.type not in ('out_invoice', 'out_refund', 'out_receipt') or not move.company_id.anglo_saxon_accounting:
+                continue
+
+            for line in move.invoice_line_ids:
+
+                # Filter out lines being not eligible for COGS.
+                if line.product_id.type != 'product' or line.product_id.valuation != 'real_time':
```

Comment:
Could you not filter the recordset instead of doing that ? The loop is long, it would reduce it a little bit